### PR TITLE
feat(core): add api for v2 of project project graph plugins

### PR DESCRIPTION
**Pull request by AgentEnder**
## Current Behavior
There are 2 methods to interact with the project graph:
- Project Inference
	- `registerProjectTargets`
	- `projectFilePatterns`
- Project graph plugins
	- 	`processProjectGraph`

These 2 methods create some pain for us as we try to work on the internal code base, and there are some issues with the current inference APIs.
- Current API requires 1:1 association between project files and projects. This doesn't work for cases like:
	- angular.json
	- someCsProject.sln
	- gradle files
- `processProjectGraph` allows adding nodes to the graph, which seems like it would work as a project inference method outside of the official inference APIs. Adding nodes like this results in the project file map not being correctly calculated though, since they are added after we calculate the map.
- We can't really optimize things for knowing when dependencies and nodes are added, since they are potentially done in parallel rather than ensuring all nodes are present first.

## Expected Behavior
Plugins have 2 APIs for updating the graph:
- `createNodes` 
- `createDependencies`

This allows us to proceed while knowing that the graph is updated with all nodes, and afterwards we can create dependencies between them.

### CreateNodes

Provides a glob pattern and factory to construct nodes from files that match that pattern. As an example, consider the implementation of the project.json node creator:

```typescript
const PROJECT_JSON_PLUGIN = {
    name: 'nx-core-build-project-json-nodes',
    createNodes: [
      '{project.json,**/project.json}',
      (file) => {
        const json = readJsonFile<ProjectConfiguration>(join(root, file));
        const project = buildProjectFromProjectJson(json, file);
        return {
          projects: {
            [project.name]: project,
          },
        };
      },
    ],
  };
```

The `createNodes` property is a tuple. The first element is the glob pattern, which says that we want all project.json files, including the one at the root if it is present. The second element is a function that takes the path to an individual file that was identified as matching the provided glob pattern, and returns all projects and external dependencies which are read from that file.

In the case of `project.json` this is a single project. A similar plugin to support `workspace.json` would look like this:

> **WARNING**
This plugin isn't included, since we don't support workspace.json any more. It is strictly an example.

```typescript
const WORKSPACE_JSON_PLUGIN = {
    name: 'nx-core-build-workspace-json-nodes',
    createNodes: [
      'workspace.json',
      (file) => {
        const json = readJsonFile(file);
        return {
          projects: json.projects
        };
      },
    ],
  };
```

### CreateDependencies

`createDependencies` returns an array of dependencies which should be added to the graph. It receives as arguments a context argument similar to the context currently provided to `processProjectGraph`

## Known Limitations

The new APIs don't include methods to remove nodes or dependencies. We can't really find a valid use case for doing so, and it would result in complicating the return types of each of the methods.

## Related Issue(s)

Fixes #

### CODE REVIEW

File: packages/nx/src/project-graph/build-nodes/workspace-projects.ts
Comment (AgentEnder) [inline packages/nx/src/project-graph/build-nodes/workspace-projects.ts:1]: We discussed previously greatly simplifying this file, but there are some mechanical differences that would arise if we move a lot of this into the project.json plugin...

Currently, with this where it is now, the normalization we do would apply to nodes which are defined by plugins. I think this is a pretty big difference, and may be a point in favor of leaving the normalization where it is. Open to opinion.

File: packages/nx/src/migrations/update-15-0-0/prefix-outputs.spec.ts [hunk 1]
```diff
@@ -83,6 +83,10 @@ describe('15.0.0 migration (prefix-outputs)', () => {
   });
 
   it('should migrate package.json projects', async () => {
+    updateJson(tree, 'package.json', (json) => {
+      json.workspaces = ['proj'];
+      return json;
+    });
```
Comment (AgentEnder) [inline packages/nx/src/migrations/update-15-0-0/prefix-outputs.spec.ts:89]: Generators used to detect all **_new_** package.json's as a project immediately. This was incorrect behavior and is addressed in the changes to project-configuration.ts

File: packages/nx/src/project-graph/build-project-graph.ts [hunk 2]
```diff
@@ -147,6 +154,9 @@ async function buildProjectGraphUsingContext(
 
   const builder = new ProjectGraphBuilder(null, ctx.fileMap);
   builder.setVersion(projectGraphVersion);
+  for (const node in knownExternalNodes) {
+    builder.addExternalNode(knownExternalNodes[node]);
+  }
 
   await buildWorkspaceProjectNodes(ctx, builder, nxJson);
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/build-project-graph.ts:161]: Please rename this function to `resolveProjectNodes`

File: packages/nx/src/project-graph/build-project-graph.ts [hunk 3]
```diff
@@ -209,13 +219,13 @@ async function updateProjectGraphWithPlugins(
...
-  const plugins = (
-    await loadNxPlugins(context.nxJsonConfiguration.plugins)
-  ).filter((x) => !!x.processProjectGraph);
+  const plugins = await loadNxPlugins(context.nxJsonConfiguration.plugins);
   let graph = initProjectGraph;
   for (const plugin of plugins) {
     try {
-      graph = await plugin.processProjectGraph(graph, context);
+      if (isNxPluginV1(plugin) && plugin.processProjectGraph) {
+        graph = await plugin.processProjectGraph(graph, context);
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/build-project-graph.ts:227]: Add a deprecation warning.

File: packages/nx/src/project-graph/build-project-graph.ts [hunk 4]
```diff
@@ -225,6 +235,25 @@ async function updateProjectGraphWithPlugins(
...
   }
+  for (const plugin of plugins) {
+    try {
+      if (isNxPluginV2(plugin)) {
+        const builder = new ProjectDependencyBuilder(graph);
+        await plugin.processProjectDependencies?.(builder, {
+          ...context,
+          graph,
+        });
+        graph = builder.getUpdatedProjectGraph();
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/build-project-graph.ts:246]: There should be another comment with my thoughts here.

File: packages/nx/src/project-graph/project-graph-builder.ts [hunk 5]
```diff
@@ -342,52 +379,80 @@ export class ProjectGraphBuilder {
...
-        }
-      }
-    }
-    return fileDeps;
+interface CandidateDependency {
+  sourceProjectName: string;
+  targetProjectName: string;
+  sourceProjectFile?: string;
+  dependencyType: DependencyType;
+}
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/project-graph-builder.ts:389]: This is a real dependency... it doesn't need to be on the graph yet.

File: packages/nx/src/project-graph/utils/retrieve-workspace-files.ts [hunk 6]
```diff
@@ -225,7 +270,7 @@ async function configurationGlobs(
 /**
  * @deprecated Use {@link configurationGlobs} instead.
  */
-function configurationGlobsSync(
+export function configurationGlobsSync(
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/utils/retrieve-workspace-files.ts:273]: Don't export this

File: packages/nx/src/config/workspaces.ts [hunk 7]
```diff
@@ -149,218 +154,142 @@ export async function getGlobPatternsFromPluginsAsync(
...
 
 export function buildProjectsConfigurationsFromProjectPaths(
   nxJson: NxJsonConfiguration,
   projectFiles: string[], // making this parameter allows devkit to pick up newly created projects
+  root: string = workspaceRoot,
   readJson: <T extends Object>(string) => T = <T extends Object>(string) =>
-    readJsonFile<T>(string) // making this an arg allows us to reuse in devkit
-): Record<string, ProjectConfiguration> {
+    readJsonFile<T>(string), // making this an arg allows us to reuse in devkit
+  _ignorePluginInference = false // When resolving a local plugin, we can't also load plugins. This would cause an infinite loop.
```
Comment (FrozenPandaz) [inline packages/nx/src/config/workspaces.ts:244]: Don't pass this flag here. Instead pass the resolved nx plugins.

File: packages/nx/src/project-graph/utils/retrieve-workspace-files.ts [hunk 8]
```diff
@@ -155,14 +186,25 @@ function buildAllWorkspaceFiles(
 function createProjectConfigurations(
   workspaceRoot: string,
   nxJson: NxJsonConfiguration,
-  configFiles: string[]
-): Record<string, ProjectConfiguration> {
+  configFiles: string[],
+  _ignorePluginInference = false
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/utils/retrieve-workspace-files.ts:190]: Split this into multiple functions, one that does not use plugins, and this one which does

File: packages/nx/src/project-graph/utils/retrieve-workspace-files.ts (hunk 6 above)
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/utils/retrieve-workspace-files.ts:273]: Don't export this

File: packages/nx/src/utils/find-matching-projects.ts [hunk 9]
```diff
@@ -245,7 +245,12 @@ export const getMatchingStringsWithCache = (() => {
     }
     const patternCache = minimatchCache.get(pattern)!;
     if (!regexCache.has(pattern)) {
-      regexCache.set(pattern, minimatch.makeRe(pattern));
+      const regex = minimatch.makeRe(pattern);
+      if (regex) {
+        regexCache.set(pattern, regex);
+      } else {
+        throw new Error('Unable to build regex for glob pattern ' + pattern);
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/find-matching-projects.ts:252]:
```suggestion
        throw new Error('Invalid glob pattern: ' + pattern);
```

File: packages/nx/src/utils/nx-plugin.ts [hunk 10]
```diff
@@ -23,31 +29,78 @@ import {
...
 import { readTsConfig } from '../plugins/js/utils/typescript';
+import { NxJsonConfiguration } from '../config/nx-json';
 
 import type * as ts from 'typescript';
 import { retrieveProjectConfigurationsWithoutPluginInference } from '../project-graph/utils/retrieve-workspace-files';
+import { NxPluginV1 } from './nx-plugin.deprecated';
+import { ProjectDependencyBuilder } from '../project-graph/project-graph-builder';
+import { combineGlobPatterns } from './globs';
+
+export type ProjectConfigurationBuilder = (
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:43]:
```suggestion
export type ProjectConfigurationConstructor = (
```

?

File: packages/nx/src/utils/nx-plugin.ts [hunk 11]
```diff
@@ -23,31 +29,78 @@ import {
...
 
   /**
-   * A glob pattern to search for non-standard project files.
-   * @example: ["*.csproj", "pom.xml"]
+   * Provides a map between file patterns and functions that retrieve configuration info from
+   * those files. e.g. { '**\/*.csproj': buildProjectsFromCsProjFile }
    */
-  projectFilePatterns?: string[];
-}
+  processProjectNodes?: Record<string, ProjectConfigurationBuilder>;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:92]:
```suggestion
  projectConfigurationsConstructors?: Record<string, ProjectConfigurationsConstructor>;
```

Consider using a tuple instead of a Record

File: packages/nx/src/utils/nx-plugin.ts [hunk 12]
```diff
@@ -23,31 +29,78 @@ import {
...
+
+  // Todo(@AgentEnder): This shouldn't be a full processor, since its only responsible for defining edges between projects. What do we want the API to be?
+  processProjectDependencies?: DependencyLocator;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:95]:
```suggestion
  projectDependencyLocator?: ProjectDependencyLocator;
```

File: packages/nx/src/utils/nx-plugin.ts [hunk 13]
```diff
@@ -174,7 +227,45 @@ export async function loadNxPlugins(
...
+}
+
+export function isNxPluginV1(plugin: NxPlugin): plugin is NxPluginV1 {
+  return (
+    ('processProjectGraph' in plugin || 'projectFilePatterns' in plugin) &&
+    !isNxPluginV2(plugin)
+  );
 }
 
 export function mergePluginTargetsWithNxTargets(
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:271]: Is this called anymore?

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/nx/src/config/workspaces.ts [hunk 14]
```diff
@@ -125,17 +126,11 @@ export function getGlobPatternsFromPlugins(
...
-    if (!plugin.projectFilePatterns) {
-      continue;
-    }
-    for (const filePattern of plugin.projectFilePatterns) {
-      patterns.push('*/**/' + filePattern);
-    }
-  }
-
-  return patterns;
+  return plugins.flatMap((plugin) =>
```
Comment (FrozenPandaz) [inline packages/nx/src/config/workspaces.ts:129]: Return this to a `for loop` we don't need a `flatMap` here

File: packages/nx/plugins/project-json.ts [hunk 15]
```diff
@@ -0,0 +1,43 @@
+import { dirname } from 'node:path';
+
+import { ProjectConfiguration } from '../src/config/workspace-json-project-json';
+import { toProjectName } from '../src/config/workspaces';
+import { readJsonFile } from '../src/utils/fileutils';
+import { NxPluginV2 } from '../src/utils/nx-plugin';
+
+export function getProjectJsonPlugin(
+  readJson: <T extends Object>(string) => T = <T extends Object>(string) =>
```
Comment (FrozenPandaz) [inline packages/nx/plugins/project-json.ts:9]: We can get rid of this now.

File: packages/nx/plugins/package-json-workspaces.ts [hunk 16]
```diff
@@ -0,0 +1,153 @@
...
+import { NX_PREFIX } from '../src/utils/logger';
+import { NxPluginV2 } from '../src/utils/nx-plugin';
+import { output } from '../src/utils/output';
+import { PackageJson } from '../src/utils/package-json';
+import { joinPathFragments } from '../src/utils/path';
+
+export function getPackageJsonWorkspacesPlugin(
+  root: string,
+  nxJson: NxJsonConfiguration,
+  readJson: <T extends Object>(string) => T
```
Comment (FrozenPandaz) [inline packages/nx/plugins/package-json-workspaces.ts:18]: We can get rid of this now.

File: packages/nx/src/project-graph/utils/retrieve-workspace-files.ts [hunk 17]
```diff
@@ -32,10 +35,16 @@ export async function retrieveWorkspaceFiles(
   workspaceRoot: string,
   nxJson: NxJsonConfiguration
 ) {
-  const { getWorkspaceFilesNative } = require('../../native');
+  const { getWorkspaceFilesNative } =
+    require('../../native') as typeof import('../../native');
 
   performance.mark('native-file-deps:start');
   let globs = await configurationGlobs(workspaceRoot, nxJson);
+  const plugins = await loadNxPlugins(
```
Comment (FrozenPandaz) [inline packages/nx/src/project-graph/utils/retrieve-workspace-files.ts:43]: configurationGlobs also loads plugins. Please ensure plugins are only loaded once.

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

File: packages/nx/src/utils/nx-plugin.ts [hunk 18]
```diff
@@ -15,39 +19,87 @@ import {
...
+  /**
+   * All files in the workspace
+   */
+  fileMap: ProjectFileMap;
+
+  /**
+   * Files changes since last invocation
+   */
+  filesToProcess: ProjectFileMap;
+}) => Awaitable<Record<string, ProjectGraphDependencyWithFile[]>>;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:82]: Inline Awaitable and return an array

File: packages/nx/src/utils/nx-plugin.ts [hunk 19]
```diff
@@ -15,39 +19,87 @@ import {
...
 
   /**
-   * A glob pattern to search for non-standard project files.
-   * @example: ["*.csproj", "pom.xml"]
+   * Provides a file pattern and function that retrieves configuration info from
+   * those files. e.g. { '**\/*.csproj': buildProjectsFromCsProjFile }
    */
-  projectFilePatterns?: string[];
-}
+  projectConfigurationsConstructor?: ProjectConfigurationsConstructor;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:91]: createNodes: CreateNodes

File: packages/nx/src/utils/nx-plugin.ts [hunk 20]
```diff
@@ -15,39 +19,87 @@ import {
...
+  projectFilePattern: string,
+  constructorMethod: (
+    projectConfigurationFile: string,
+    context: {
+      projectsConfigurations: Record<string, ProjectConfiguration>;
+      nxJsonConfiguration: NxJsonConfiguration;
+      workspaceRoot: string;
+    }
+  ) => {
+    projectNodes?: Record<string, ProjectConfiguration>;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:51]: The property is projectNodes but the type is project configurations. The name should align?

File: packages/nx/src/utils/nx-plugin.ts [hunk 21]
```diff
@@ -15,39 +19,87 @@ import {
...
-   * @example: ["*.csproj", "pom.xml"]
+   * Provides a file pattern and function that retrieves configuration info from
+   * those files. e.g. { '**\/*.csproj': buildProjectsFromCsProjFile }
    */
-  projectFilePatterns?: string[];
-}
+  projectConfigurationsConstructor?: ProjectConfigurationsConstructor;
+
+  // Todo(@AgentEnder): This shouldn't be a full processor, since its only responsible for defining edges between projects. What do we want the API to be?
+  projectDependencyLocator?: ProjectDependencyLocator;
```
Comment (FrozenPandaz) [inline packages/nx/src/utils/nx-plugin.ts:94]: createDependencies: CreateDependencies

**Review by FrozenPandaz — CHANGES_REQUESTED** [review]

**Review by Coly010 — APPROVED** [review]
The file that I've been tagged to review looks good

**Review by FrozenPandaz — APPROVED** [review]

File: typedoc-theme/src/lib/theme.ts [hunk 22]
```diff
@@ -54,14 +54,14 @@ export default class NxMarkdownTheme extends MarkdownTheme {
         directory: '.',
         template: this.getReflectionTemplate(),
       },
+      {
+        kind: [ReflectionKind.TypeAlias],
+        isLeaf: true,
+        directory: '.',
+        template: this.getReflectionMemberTemplate(),
+      },
```
Comment (isaacplmann) [inline typedoc-theme/src/lib/theme.ts:62]: Why was this moved?  I don't think it makes a difference in the output.

**Review by isaacplmann — CHANGES_REQUESTED** [review]

File: typedoc-theme/src/lib/theme.ts (hunk 22 above)
Comment (AgentEnder) [reply typedoc-theme/src/lib/theme.ts]: Huh, I would have sworn it did, but removing it and running "nx documentation" resulted in no changes. I'll revert it.

**Review by isaacplmann — APPROVED** [review]

**Review by leosvelperez — APPROVED** [review]

**Review by meeroslav — APPROVED** [review]
